proofs/pfedit.ml

```diff
@@ -101,6 +101,12 @@ let solve ?with_end_tac gi info_lvl tac pr =
       | Some _ -> Proofview.Trace.record_info_trace tac
     in
     let tac = match gi with
+      | Vernacexpr.SelectAlreadyFocused ->
+        let open Proofview.Notations in
+        Proofview.numgoals >>= fun n ->
+        if n == 1 then tac
+        else CErrors.user_err Pp.(str "Expected a single focused goal but " ++
+                                  int n ++ str " goals are focused.")
```

Comment:
Even though observationally equivalent in this context, I'd rather use `tclZEROMSG` rather than raising an exception here, it'd be safer if ever the surrounding code gets modified.

Reply:
>Even though observationally equivalent in this context, I'd rather use tclZEROMSG rather than raising an exception here, it'd be safer if ever the surrounding code gets modified.

OK but can you explain more about how it's safer?

Reply:
This currently works because the only thing you do with the resulting monadic value is to run it. But if you were to compose it with other code, due to the very higher-orderedness of the monad, the exception would get raised at an unpredictible time and would most likely wreak havoc w.r.t. intended semantics. Except for very specific cases that are documented (e.g. `Goal.enter`) one should not raise exceptions inside the monadic layer.

Reply:
ZEROMSG is not available in pfedit x_x

Reply:
Well, you can always use `tclZERO` + `UserError`...

Reply:
>Well, you can always use tclZERO + UserError...

That's what I did.

Reply:
No, you used `user_error` which is just a wrapper raising an exception. You should use a monad combinator instead.

Reply:
The PR has been updated since then.